### PR TITLE
PR: Fix running namespace and improve eventloop integration while debugging

### PR DESCRIPTION
**Pull request by impact27**
Fixes https://github.com/spyder-ide/spyder-kernels/issues/394

### CODE REVIEW

File: spyder_kernels/customize/spyderpdb.py [hunk 1]
```diff
@@ -100,6 +100,8 @@ def __init__(self, completekey='tab', stdin=None, stdout=None,
         # has no effect in previous versions.
         self.report_skipped = False
 
+        self.executing = False
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:103]: Please add a comment on top of this variable to explain why it's necessary or how it's used.

**Review by ccordoba12 — COMMENTED** [review]
@impact27, I left a tiny comment for you, otherwise this looks good to me.

File: spyder_kernels/customize/spyderpdb.py (hunk 1 above)
Comment (impact27) [reply spyder_kernels/customize/spyderpdb.py]: Slight change: What I am actually interested in is to know if a pdb prompt is waiting. I adjusted the code and moved `cmd_input` which should have been in `SpyderPdb`

File: spyder_kernels/customize/spydercustomize.py [hunk 2]
```diff
@@ -328,18 +327,8 @@ def _patched_get_terminal_size(fd=None):
...
-def cmd_input(prompt=''):
-    return get_ipython().kernel.cmd_input(prompt)
-
-
 pdb.Pdb = SpyderPdb
 
-if PY2:
-    cmd.raw_input = cmd_input
-else:
-    cmd.input = cmd_input
```
Comment (impact27) [inline spyder_kernels/customize/spydercustomize.py:340]: That was hacky, I am glad to have found a better way

File: spyder_kernels/customize/spyderpdb.py [hunk 3]
```diff
@@ -99,7 +101,14 @@ def __init__(self, completekey='tab', stdin=None, stdout=None,
         # has no effect in previous versions.
         self.report_skipped = False
 
-        self.executing = False
+        # Keep track of remote filename
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:104]:
```suggestion

        # Keep track of remote filename
```

File: spyder_kernels/customize/spyderpdb.py [hunk 4]
```diff
@@ -99,7 +101,14 @@ def __init__(self, completekey='tab', stdin=None, stdout=None,
...
+        self.remote_filename = None
+
+        # State of the prompt
+        self.prompt_waiting = False
+
+        # Line recieved from the frontend
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:110]:
```suggestion
        # Line received from the frontend
```

File: spyder_kernels/customize/spyderpdb.py [hunk 5]
```diff
@@ -636,14 +645,82 @@ def _cmdloop(self):
                        "For copying text while debugging, use Ctrl+Shift+C",
                        file=self.stdout)
 
+    def cmdloop(self, intro=None):
+        """Repeatedly issue a prompt, accept input, parse an initial prefix
+        off the received input, and dispatch to action methods, passing them
+        the remainder of the line as argument.
+
+        """
+
```
Comment (ccordoba12) [inline spyder_kernels/customize/spyderpdb.py:650]:
```suggestion

    def cmdloop(self, intro=None):
        """
        Repeatedly issue a prompt, accept input, parse an initial prefix
        off the received input, and dispatch to action methods, passing them
        the remainder of the line as argument.
        """
```

**Comment by ccordoba12** [comment]
This looks good to me now @impact27. As before, I'd like to ask you: do you think it's worth it to include these changes in 2.3.2 ? Considering stability in mind, of course.

**Comment by impact27** [comment]
The bug is quite bad because the namespace is wrong while debigging outside of the main file. I think the risk is relatively low, because the tests now cover this part better.

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27!

**Pull request merged**